pkg/virtualKubelet/reflection/storage/persistentvolumeclaim.go

```diff
@@ -0,0 +1,182 @@
+
+			provisionerName: consts.StorageProvisionerName,
+
+			remoteClusterID:            remoteClusterID,
+			virtualStorageClassName:    virtualStorageClassName,
+			remoteRealStorageClassName: remoteRealStorageClassName,
+
+			eventRecorder: eventRecorder,
+		}
+	}
+}
+
+// Handle reconciles PersistentVolumeClaim objects.
+func (npvcr *NamespacedPersistentVolumeClaimReflector) Handle(ctx context.Context, name string) error {
+	// Retrieve the local and remote objects (only not found errors can occur).
+	klog.V(4).Infof("Handling reflection of local PersistentVolumeClaim %q (remote: %q)", npvcr.LocalRef(name), npvcr.RemoteRef(name))
+	local, lerr := npvcr.localPersistentVolumeClaims.Get(name)
+	utilruntime.Must(client.IgnoreNotFound(lerr))
+	remote, rerr := npvcr.remotePersistentVolumeClaims.Get(name)
+	utilruntime.Must(client.IgnoreNotFound(rerr))
+
+	// Abort the reflection if the remote object is not managed by us, as we do not want to mutate others' objects.
+	if rerr == nil && !forge.IsReflected(remote) {
+		klog.Infof("Skipping reflection of local PersistentVolumeClaim %q as remote already exists and is not managed by us", npvcr.LocalRef(name))
+		return nil
+	}
+
+	// The local PersistentVolumeClaim does no longer exist. Ensure it is also absent from the remote cluster.
+	if kerrors.IsNotFound(lerr) {
+		if !kerrors.IsNotFound(rerr) {
+			klog.V(4).Infof("Deleting remote PersistentVolumeClaim %q, since local %q does no longer exist", npvcr.RemoteRef(name), npvcr.LocalRef(name))
+			return npvcr.DeleteRemote(ctx, npvcr.remotePersistentVolumesClaimsClient, PersistentVolumeClaimReflectorName, name, remote.GetUID())
+		}
+
+		klog.V(4).Infof("Local PersistentVolumeClaim %q and remote PersistentVolumeClaim %q both vanished", npvcr.LocalRef(name), npvcr.RemoteRef(name))
+		return nil
+	}
+
+	if should, err := npvcr.shouldProvision(local); err != nil {
+		klog.V(4).Infof("Error checking if should provision a local PersistentVolumeClaim %q: %q", npvcr.LocalRef(name), err.Error())
+		return err
+	} else if !should {
+		klog.V(4).Infof("Skipping PersistentVolumeClaim %q since we should not provision it", npvcr.LocalRef(name))
+		return nil
+	}
+
+	state, err := npvcr.provisionClaimOperation(ctx, local,
+		func(c context.Context, options controller.ProvisionOptions) (*v1.PersistentVolume, controller.ProvisioningState, error) {
+			if clusterID, found := utils.GetNodeClusterID(options.SelectedNode); !found || clusterID != npvcr.remoteClusterID {
+				return nil, controller.ProvisioningFinished, &controller.IgnoredError{Reason: "this provisioner is not provisioning storage on that node"}
+			}
+
+			pv, state, err := liqostorageprovisioner.ProvisionRemotePVC(ctx,
+				options, npvcr.RemoteNamespace(), npvcr.remoteRealStorageClassName,
+				npvcr.remotePersistentVolumeClaims, npvcr.remotePersistentVolumesClaimsClient)
+			if err == nil && state == controller.ProvisioningFinished {
+				local.Spec.VolumeName = options.PVName
```

Comment:
You cannot mutate the objects retrieved from the lister without deepcopying, otherwise you mutate the cache.